core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1662,11 +1785,46 @@ public Optional<NodeRequirements> getNodeRequirements(int partitionId)
             return getStagePartition(partitionId).getNodeRequirements();
         }
 
-        public OutputDataSizeEstimate getOutputDataSize()
+        public OutputDataSizeEstimateResult getOutputDataSize()
         {
-            // TODO enable speculative execution
-            checkState(stage.getState() == StageState.FINISHED, "stage %s is expected to be in FINISHED state, got %s", stage.getStageId(), stage.getState());
-            return new OutputDataSizeEstimate(ImmutableLongArray.copyOf(outputDataSize));
+            if (stage.getState() == StageState.FINISHED) {
+                return OutputDataSizeEstimateResult.complete(new OutputDataSizeEstimate(ImmutableLongArray.copyOf(outputDataSize)));
+            }
+            if (isNoMorePartitions()) {
+                return OutputDataSizeEstimateResult.unknown();
+            }
+
+            int allPartitionsCount = getPartitionsCount();
+            int remainingPartitionsCount = getRemainingPartitionsCount();
+
+            if (remainingPartitionsCount == allPartitionsCount) {
+                return OutputDataSizeEstimateResult.unknown();
+            }
+
+            double extrapolationRatio = (double) (allPartitionsCount - remainingPartitionsCount) / allPartitionsCount;
+            ImmutableLongArray.Builder estimateBuilder = ImmutableLongArray.builder(outputDataSize.length);
+
+            for (long partitionSize : outputDataSize) {
+                estimateBuilder.add((long) (partitionSize / extrapolationRatio));
+            }
+
+            return OutputDataSizeEstimateResult.partial(
+                    new OutputDataSizeEstimate(estimateBuilder.build()),
+                    extrapolationRatio);
+        }
+
+        public double getOutputDataSizeConfidence()
+        {
+            if (stage.getState() == StageState.FINISHED) {
+                return 1.0;
+            }
+            if (isNoMorePartitions()) {
+                return 0.0;
```

Comment:
Why does a stage having no more partitions mean our confidence is zero? Maybe i don't understand when `noMorePartitions()` is called, but that _sounds_ to me like it means the stage will have not create any additional partitions. Is that wrong?

Reply:
it is missing `!`. Should be `!sNoMorePartitions()` - which would mean "we did not enumerate how many tasks will there be for a stage so we cannot estimate"